### PR TITLE
TT-7216 Decode Option For Raw Request/Response

### PR DESCRIPTION
**Pull request by Firaasss**
## Description
Added an option to decode the raw request and raw response data of each analytical record from base64 prior to writing them to the data sink. This solution is implemented at the same level as other general pump filters. 

A configuration option has been included to allow the decoding feature to be turned on or off (off by default) in `config.go`.  

## Related Issue
https://tyktech.atlassian.net/browse/TT-7216

## Motivation and Context
This gives additional functionality to our Tyk-Pump component to allow the encoded transactional logs to be decoded for human readability. This ticket covers two previous requests (from community and customers) that were specifically for syslog and csv. Both of these reference the fact that this functionality is already implemented for the Elasticsearch pump.

## How This Has Been Tested
There are a variety of tests cases that are made in `main_test.go` to generate scenarios that will test decoding the raw request/response in 4 different scenarios. 

Another test using an stdout pump while leveraging the pump demo (using --demo tag) was done to test if the values from the config are actually working. This was done using a redis and mongo service to verify the values that were being written after setting the config value and decoding the raw response/request. 

Eg. Setting `"raw_request_decoded": false` and `"raw_response_decoded": true` in `pump.conf`

Screen Shot 2023-01-05 at 1 15 05 PM

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [x] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [x] My change requires a change to the documentation.
  - [x] If you've changed APIs, describe what needs to be updated in the documentation.
- [ ] I have updated the documentation accordingly.
- [x] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -300,7 +302,10 @@ func filterData(pump pumps.Pump, keys []interface{}) []interface{} {
 	shouldTrim := SystemConfig.MaxRecordSize != 0 || pump.GetMaxRecordSize() != 0
 	filters := pump.GetFilters()
 	ignoreFields := pump.GetIgnoreFields()
-	if !filters.HasFilter() && !pump.GetOmitDetailedRecording() && !shouldTrim && len(ignoreFields) == 0 {
+	getDecodingResponse := pump.GetDecodedResponse()
+	getDecodingRequest := pump.GetDecodedRequest()
+	if !getDecodingRequest && !getDecodingResponse && !filters.HasFilter() && !pump.GetOmitDetailedRecording() && !shouldTrim && len(ignoreFields) == 0 {
+		fmt.Println("Returning here")
```
Comment (sredxny) [inline main.go:308]: please remove this debug line

**Review by sredxny — CHANGES_REQUESTED** [review]
lgtm but requested a small change

**Review by mativm02 — APPROVED** [review]
well done @Firaasss! please, remove that testing print :) overall, an amazing job!

File: main.go [hunk 2]
```diff
@@ -300,7 +302,10 @@ func filterData(pump pumps.Pump, keys []interface{}) []interface{} {
 	shouldTrim := SystemConfig.MaxRecordSize != 0 || pump.GetMaxRecordSize() != 0
 	filters := pump.GetFilters()
 	ignoreFields := pump.GetIgnoreFields()
-	if !filters.HasFilter() && !pump.GetOmitDetailedRecording() && !shouldTrim && len(ignoreFields) == 0 {
+	getDecodingResponse := pump.GetDecodedResponse()
+	getDecodingRequest := pump.GetDecodedRequest()
+	if !getDecodingRequest && !getDecodingResponse && !filters.HasFilter() && !pump.GetOmitDetailedRecording() && !shouldTrim && len(ignoreFields) == 0 {
```
Comment (sredxny) [inline main.go:307]: also, could you explain the logic in this conditional? it can be an inline comment or something, as it's a pretty long conditional

**Review by sredxny — DISMISSED** [review]

File: main.go (hunk 1 above)
Comment (Firaasss) [reply main.go]: Removed and pushed.

File: main.go (hunk 2 above)
Comment (Firaasss) [reply main.go:308]: Yes done - my addition to this code was `!getDecodingRequest `and `!getDecodingResponse` to check if the configs for both of those was false. Essentially the whole conditional is to see if all config options are empty or false.

**Pull request merged**